src/server.cpp

```diff
@@ -168,138 +171,216 @@ void installSignalHandlers() {
-
-        int workers = config.restWorkers() ? config.restWorkers() : 10;
-        SPDLOG_INFO("Will start {} REST workers", workers);
-
-        std::unique_ptr<ovms::http_server> restServer = ovms::createAndStartHttpServer(config.restBindAddress(), config.restPort(), workers);
-        if (restServer != nullptr) {
-            SPDLOG_INFO("Started REST server at {}", server_address);
-        } else {
-            throw std::runtime_error("Failed to start REST server at " + server_address);
+};
+class HTTPServerModule : public Module {
+    std::unique_ptr<ovms::http_server> server;
+
+public:
+    HTTPServerModule() = default;
+    int start(const ovms::Config& config) override {
+        if (config.restPort() != 0) {
+            state = ModuleState::STARTED_INITIALIZE;
+            const std::string server_address = config.restBindAddress() + ":" + std::to_string(config.restPort());
+            int workers = config.restWorkers() ? config.restWorkers() : 10;
+
+            SPDLOG_INFO("Will start {} REST workers", workers);
+            server = ovms::createAndStartHttpServer(config.restBindAddress(), config.restPort(), workers);
+            if (server != nullptr) {
+                SPDLOG_INFO("Started REST server at {}", server_address);
+            } else {
+                SPDLOG_ERROR("Failed to start REST server at " + server_address);
+                return EXIT_FAILURE;
+            }
+            state = ModuleState::INITIALIZED;
+        }
+        return EXIT_SUCCESS;
+    }
+    void shutdown() override {
+        if (server != nullptr) {
+            state = ModuleState::STARTED_SHUTDOWN;
+            server->Terminate();
+            server->WaitForTermination();
+            SPDLOG_INFO("Shutdown HTTP server");
+            state = ModuleState::SHUTDOWN;
         }
-
-        return restServer;
     }
+};
 
-    return nullptr;
+ModuleState Server::getModuleState(const std::string& name) const {
+    auto it = modules.find(name);
+    if (it == modules.end())
+        return ModuleState::NOT_INITIALIZED;
+    return it->second->getState();
 }
 
-int server_main(int argc, char** argv) {
-    installSignalHandlers();
-    try {
-        auto& config = ovms::Config::instance().parse(argc, argv);
-        configure_logger(config.logLevel(), config.logPath());
+class ProfilerModule : public Module {
+    std::unique_ptr<Profiler> profiler;
 
+public:
+    ProfilerModule() = default;
+    int start(const Config& config) override {
 #ifdef MTR_ENABLED
-        auto profiler = Profiler(config.tracePath());
+        state = ModuleState::STARTED_INITIALIZE;
+        auto profiler = std::make_unique<Profiler>(config.tracePath());
         if (!profiler.isInitialized()) {
             SPDLOG_ERROR("Cannot open file for profiler, --trace_path: {}", config.tracePath());
             return EXIT_FAILURE;
         }
+        state = ModuleState::INITIALIZED;
 #endif
+        return EXIT_SUCCESS;
+    }
+    void shutdown() override {
```

Comment:
Here you can shutdown non initialized module. (When compiled without MTR_ENABLED flag)